src/accounts/Market.ts

```
@@ -1,29 +1,249 @@
+		const investorLPTokenAccount = await this.findLPTokenAccount(investor);
+		const [credixPass] = await this.generateCredixPassPDA(investor);
+
+		if (!gatewayToken) {
+			throw Error("No valid Civic gateway token found");
+		}
+
+		return this.program.rpc.withdrawFunds(new BN(amount.toNumber()), {
+			accounts: {
+				investor,
+				gatewayToken: gatewayToken.publicKey,
+				globalMarketState: this.address,
+				signingAuthority: signingAuthority,
+				investorLpTokenAccount: investorLPTokenAccount,
+				investorTokenAccount: investorTokenAccount,
+				liquidityPoolTokenAccount: liquidityPoolTokenAccount,
+				treasuryPoolTokenAccount: this.treasury,
+				lpTokenMintAccount: this.lpMintPK,
+				credixPass,
+				baseMintAccount: this.baseMintPK,
+				tokenProgram: TOKEN_PROGRAM_ID,
+				associatedTokenProgram: ASSOCIATED_TOKEN_PROGRAM_ID,
+			},
+		});
+	}
+
+	async createDeal(
+		principal: Big,
+		financingFee: number,
+		timeToMaturity: number,
+		borrower: PublicKey,
+		dealName: string
+	) {
+		// TODO: add validations
+		if (timeToMaturity % 30) {
+			throw Error("Time to maturity needs to be a multiple of 30");
+		}
+
+		const [borrowerInfoAddress, borrowerInfoBump] = await BorrowerInfo.generatePDA(borrower, this);
+		const borrowerInfo = await this.fetchBorrowerInfo(borrower);
+		const dealNumber = borrowerInfo ? borrowerInfo.numberOfDeals + 1 : 0;
+		const [dealAddress, dealBump] = await Deal.generatePDA(borrower, dealNumber, this);
+		const [credixPassAddress] = await CredixPass.generatePDA(borrower, this);
+		const gatewayToken = await this.client.getGatewayToken(borrower, this.gateKeeperNetwork);
+
+		const principalAmount = new BN(principal.toNumber());
+		// TODO: do we need this dependency?
+		const financingFreeFraction = new Fraction(financingFee);
+		const financingFeeRatio = new Ratio(financingFreeFraction.n, financingFreeFraction.d * 100);
+
+		if (!gatewayToken) {
+			throw new Error("No valid Civic gateway token found");
+		}
```

We're doing a lot of these gateway token checks after we've done a bunch of async calls. Maybe it's better to throw these errors at the top of the function?